src/Main.elm

```diff
@@ -9,89 +9,151 @@ import Platform.Cmd as Cmd
 
 
-lightStateToWebSocketMsg : LightState -> String
-lightStateToWebSocketMsg lightState =
-    case lightState of
-        LightOn ->
-            "STATE_ON"
+stringToLightState : String -> Bool
+stringToLightState str =
+    case str of
+        "STATE_ON" ->
+            True
 
-        LightOff ->
-            "STATE_OFF"
+        "STATE_OFF" ->
+            False
 
+        _ ->
+            False
 
-lightStateToStr : LightState -> String
-lightStateToStr lightState =
-    case lightState of
-        LightOn ->
-            "ON"
 
-        LightOff ->
-            "OFF"
+lightStateToWebSocketMsg : Bool -> String
+lightStateToWebSocketMsg lightState =
+    if lightState then
+        "STATE_ON"
+
+    else
+        "STATE_OFF"
 
 
-toggleLightSwitch : LightState -> LightState
-toggleLightSwitch lightState =
-    case lightState of
-        LightOn ->
-            LightOff
+lightStateToStr : Bool -> String
+lightStateToStr lightState =
+    if lightState then
+        "ON"
 
-        LightOff ->
-            LightOn
+    else
+        "OFF"
 
 
 type Msg
-    = LightSwitched LightState
+    = LightSwitched Bool
     | WebSocketGotMsg String
     | ColourChange ColourPicker.Colour String
 
 
-initModel : Model
-initModel =
-    { lightState = LightOff
-    , rgba = ColourPicker.initRgba
-    }
-
-
 init : () -> ( Model, Cmd Msg )
 init _ =
-    ( initModel, Cmd.none )
+    ( Syncing NotSynced, Cmd.none )
 
 
 update : Msg -> Model -> ( Model, Cmd.Cmd Msg )
 update msg model =
     case msg of
-        LightSwitched state ->
-            ( { model | lightState = state }
+        WebSocketGotMsg string ->
+            ( updateStateFromWebsocketMsg string model, Cmd.none )
+
+        _ ->
+            case model of
+                Syncing _ ->
+                    ( model, Cmd.none )
+
+                Connected state ->
+                    state
+                        |> updateState msg
+                        |> Tuple.mapFirst Connected
+
+
+updateState : Msg -> State -> ( State, Cmd.Cmd Msg )
+updateState msg state =
+    case msg of
+        LightSwitched lightState ->
+            ( { state | lightState = lightState }
             , Ports.outgoingWebsocketMsg
-                (lightStateToWebSocketMsg state)
+                (lightStateToWebSocketMsg lightState)
             )
 
-        WebSocketGotMsg string ->
-            ( updateModelFromWebsocketMsg string model, Cmd.none )
-
         ColourChange colourType colour ->
             ( ColourPicker.updateColour
                 colourType
                 colour
-                model
+                state
             , Ports.outgoingWebsocketMsg
                 (ColourPicker.colourToString colour colourType)
             )
 
+        _ ->
+            ( state, Cmd.none )
+
+
+updateFromSyncingStatus : String -> Maybe String -> SyncingOnConnect -> SyncingOnConnect
+updateFromSyncingStatus cmd value syncingStatus =
+    case ( syncingStatus, value ) of
```

Comment:
We're effectively checking three values for each branch here: `syncingStatus`, `value` and ` cmd`. 
- Could use a three-tuple in the case statement (I think Elm lets us go up to 3?)

In each case the check on `val` is the same: is it Just?  Whereas the check on `cmd` is more interesting, and relates to the `syncingStatus`. So an alternative would be to wrap the whole thing in a `Maybe.map` to get the `val` value from `value` and then use a Pair in your case statement of `syncingStatus` and `cmd`?

Reply:
ooh nice, I just didn't think of that. Thanks!